### ISSUE TITLE
Redesign Selection-Keyboard.

### ISSUE THREAD

**Issue by Glitchy-Tozier** [florisboard/florisboard#216]
Selection-Keyboard:
![Screenshot_20210120-141443_GitHub.jpg](https://user-images.githubusercontent.com/59611881/105179858-2310f500-5b2a-11eb-8071-b4076bed6f01.jpg)

As of now, I feel like there is almost no reason to ever use the selection-keyboard. Especially with how well the copy&paste-bar was implemented.
![Screenshot_20210120-141423_GitHub.jpg](https://user-images.githubusercontent.com/59611881/105179961-42a81d80-5b2a-11eb-80b3-4eac9b6cdb2d.jpg)

**I also have no idea how to change it to make it viable.**
What I can offer though are some small improvements. They still won't make me think it is usable, but they might be helpful to other people.
1. **Add Delete-Swipe** to the backspace key.
2. **Add Gestures** to the selection-keyboard. (The same ones seen in the regular keyboard)
3. **The space above the selection-keyboard seems like a waste of space.** Think of what could be useful there (We _could_ put the same bar there that is present in all the other keyboards, but i don't like that thought. It seems redundant.(As the selection-keyboard already contains much of those functionalities)
4. **Make the buttons prettier.** currently it just looks like it's lines on a background.
5. Those **up&down arrow-keys** are great!
But why are they disabled when activating the "select text"-mode? This is when they are needed the most!!
6. **Two Buttons that can be added:** `Go to end of line` and `Go to beginning of Line`. I'm not talking about "go to the beginning/end of the whole document".
What I mean is "Move the cursor only to the beginning/end of the current line.

**Comment by patrickgold** [comment]
Thanks for your detailed suggestion for improving the selection keyboard!

IMO the selection keyboard is not useful if you are in a simple one-liner text field, but in a big document where navigating may be more difficult. For single-liners the clipboard toolbar is fully sufficient.

The main reason there's even a selection keyboard in FlorisBoard is because Gboard has one and if someone is used to using this selection keyboard, then they probably want it in FlorisBoard too. Though I have to admit the last time I was in the selection keyboard was when I added it, so yeah - in general, the clipboard cursor row is more useful, as accessing it is much quicker.

Now to your numbered bullet points:

1. Can definitely be added and would be useful.
2. The selection keyboard uses an entire different touch logic, so this would require some work to port it from the KeyboardView while not providing that much usefulness IMO. Maybe I will add this, I will see.
3. Well what should I show in there - the clipboard row is completely redundant here and the suggestion strip also makes no sense. Still I want to avoid hiding the Smartbar as this causes the keyboard height to change and the target app will have to recalculate & redraw its content, where there's a possibility for content-jump, which is bad UX.
4. It was intended that it is lines on a background (ok granted I copied Gboard's style here xd) but I personally wound find it ugly if it was the design of the normal keys but in big.
5. This actually has the reason to provide a better UX, because I couldn't get the selection to properly expand how one would expect it while in selection mode. So I just disabled the up/down arrow while a selection is active to avoid strange results.
6. This can be done when using the text fields characters and searching for the next `\n` (new line character). Currently, I just send a hardware-emulated key to go to the start/end, which seems to go to the start/end of a document.

I have thought about maybe completely rewriting the selection keyboard (so it uses KeyboardView as a base, thus having support for gestures), but I am out of ideas here. I will place this issue on my todo list for v0.5.0 for now, maybe I gather some ideas until then.

**Comment by Glitchy-Tozier** [comment]
I must say though that if the main purpose of this is to keep Gboard compatibility, it's definately not that important to invest a lot of time into it.
Especially with that amazing copy&paste-smart-bar.  :)

Stuff like adding word suggestions or fixing ugly bugs (like the short selection-board pop-up) seem  way more important to me.

**Comment by patrickgold** [comment]
> I must say though that if the main purpose of this is to keep Gboard compatibility, it's definately not that important to invest a lot of time into it.

Yeah that's why it is on the v0.5.0 milestone :)

>  Especially with that amazing copy&paste-smart-bar. :)

:)

> Stuff like adding word suggestions or fixing ugly bugs (like the short selection-board pop-up) seem way more important to me.

Definitely. v0.3.4, which will be released tomorrow, is basically a bug fix release (e.g. it fixes the super annoying double press issue for fast writers).

Word suggestions will be my next main dev goal for sure, alongside with importing dictionaries dynamically (import from file system). I am unsure if I make some "alpha alpha GitHub only releases" so interested persons can try this out while it is in development or if I just let people use the auto-generated debug artifacts (if they want to test out each new step). Especially at the beginning it will be very interesting to see how I implement suggestions and things in this area may change from commit to commit. My plan for supported languages is `English (US)` only until the basic structure and code for suggestions is there. Then I will slowly expand it to `Portuguese [(Brazil)]` and `German`, and if that works further and further until I have all languages that the dictionary source supports.

**Comment by Glitchy-Tozier** [comment]
German's what I'm looming for, so I'm happy!  :)

Also, I'm kind of afraid of downloading stuff outside of F-droid/Play, but I'll do my best to give feedback on the versions that get through to me!

**Comment by patrickgold** [comment]
Yes, updates via Google Play take about 2-3 days until they arrive at the user from the point I upload it to Google Play, while in F-Droid it seems to be longer (as they build the app from source and their resources are more limited than Google's).

**Comment by Glitchy-Tozier** [comment]
Switching to the Play-version then!

**Comment by yashpalgoyal1304** [comment]
I hope the word suggestions will be additional things and not bundled in. 'cause the app size is a big premium for me. Bundling the dictionaries in will increase the size a lot (i guess), and i dont want to ditch another sooo cool app. (i already have to ditch open keyboard as its app size jumped from 24 MiB {1.3} to 40 MiB {1.4.2}

Please please please dont fall for word suggestions/dictionaries. I am sure there would be some external package for dictionaries - that people can download externally - considering this is such a common thing. Ur keyboard is one of my main keyboards (and i use F-Droid only keyboard apps) and i cant afford to ditch it.

**Comment by patrickgold** [comment]
@yashpalgoyal1304 Well how should I put it - it is very difficult to find a way to solve this problem without disagreeing with at least 1 side. I've got quite some feedback about this. Some people told me that they hate the way AnySoftKeyboard handles dictionaries because for everything you need you basically have to visit the app store to install it afterwards and they want to have it bundled. Others have told me that it is important that the app size stays small (including you) and that they want to be able to load dictionaries after installation.

Now there's a very elegant solution to this, it it wasn't for this: FlorisBoard is a privacy-focused keyboard and thus I've set the goal to NEVER use the `NETWORK` permission in the core app. This disables the else beautiful solution to provide a dynamic in-app "dictionary-store" which would give you ability to select and download additional dictionaries (hosted either on GitHub or some CDN) without ever leaving the FlorisBoard app.

As for the v0.3.8 release, which will wrap up Phase 1 of the suggestion implementation, the only dictionary that will be bundled in is the English (US) one and it only has a size of ~1.7MB.

For Phase 2, I will definitely look into this issue though and hopefully find a way to solve this problem while avoiding the use of the `NETWORK` permission. My ideas:

1. Solve it how AnySoftKeyboard did, with dictionary packs on GPlay and F-Droid (Personally I dislike this idea though)
2. Make a beautiful little web page which gives you the ability to download a dictionary file form the GitHub repo. Then you either go to FlorisBoard and open the file selector, select the file and import it. Or you share the downloaded file with FlorisBoard via the Android share menu. This solution is way less elegant than the inbuilt dictionary store, though way more privacy friendly.
3. ....

What are your thoughts to my ideas? If you have another possible idea, feel free to suggest it to me.

**Comment by Glitchy-Tozier** [comment]
1. Not sure whether this is the right place to discuss this topic 😅
2. As someone who doesn't have to battle with storage: I think this is such an essential feature, that it should be implemented into the app by default.
Also, it's the one feature that non tech-savvy people will need, and making them jump through hoops may not be optimal.
(Compare that to gifs, whose target audience will be a lot younger)

That being said, adding tons of dictionaries isn't perfect either, so no idea what to suggest.

**Comment by yashpalgoyal1304** [comment]
thoughts:
- Separate app for each language's dictionary pack doesnt seem right.
- But having a single app for all the managing all of these seems like a good way to go.
- This can be extended to house all non-bare bones things - dictionaries, and gifs (mentioned in above comment by glitchy)
- Maybe something like global dictionaries can be figured out\
which can be used by multiple keyboards system wide (just like libraries i guess). It sure seems a bit of too much of a hassle in the beginning, but i think it can save time and space. If Keyboard devs like u, **simple board**, **open board** etc join forces to tackle dictionary tasks. Be it some file based, or app based.

**Comment by yashpalgoyal1304** [comment]
ohh, maybe #109 is more suited for this kind of discussion?? i will be more than happy to cut paste my input over there to keep this issue clean.

**Comment by patrickgold** [comment]
@yashpalgoyal1304 I agree this issue's topic does not match the current discussion. I would suggest moving the discussion to a new issue.

**Comment by yashpalgoyal1304** [comment]
okay. new issue it is then. i was checking for duplicates and thought that #109 could be an existing one.

**Comment by yashpalgoyal1304** [comment]
@patrickgold @Glitchy-Tozier please cut paste ur comments there ;)
{i would do it myself by quoting, but then u wont be able to edit that. so...}

**Comment by MisterY** [comment]
Select All is missing here.

**Comment by yashpalgoyal1304** [comment]
it is shown to me, the first item in right corner:

![zjw3NnJ7ec](https://user-images.githubusercontent.com/19423063/107872959-6ca7f200-6ed4-11eb-89cf-5643db8d680d.png)

**Comment by MisterY** [comment]
Which version are you on? This is 0.3.7, the latest available on F-Droid and GitHub

![Screenshot_20210220-194556_Firefox_Beta](https://user-images.githubusercontent.com/462445/108605442-5f977100-73b4-11eb-901f-fa112428987f.png)

**Comment by MisterY** [comment]
There's probably no need to keep the Gboard layout. The cursor keys may be arranged as on a PC keyboard, leaving room for a whole row of other keys, like cut/copy/paste with icons. The right side may be Select Text, Select All, etc.

**Comment by yashpalgoyal1304** [comment]
> or something like <^v>

wow, that'd be nice. but now that it's discussed, these can be in the order what world already recognises: <kbd><v^></kbd> i.e. ◀🔽🔼▶\
ref: 
- https://en.wikipedia.org/wiki/Arrow_keys#HJKL_keys
- https://www.ecosia.org/search?tt=0e30b125&q=hjkl%20cursor%20keys

**Comment by Glitchy-Tozier** [comment]
> There's probably no need to keep the Gboard layout.

As far as I know, the selection-board was only added to give Gboard users the thing they're used to. Meaning there's some inclination to keep it gboard-compatible.

**Comment by pokegamer5547** [comment]
I'm one of the fewest person who actually uses gboard selection keyboard and it prefers having it atleast as a option. It helps me quite a bit when writing notes in note taking apps. And also i prefer using just one hand mostly. Will this feature gonna be reimplemented again in future?

**Comment by patrickgold** [comment]
> Will this feature gonna be reimplemented again in future?

Yes, definitely! But more feature packed than Gboard's idea, leaning towards Hacker's keyboard probably (so you have all meta keys, arrow keys and more in the keyboard at once).

**Comment by pokegamer5547** [comment]
> > Will this feature gonna be reimplemented again in future?
> 
> Yes, definitely! But more feature packed than Gboard's idea, leaning towards Hacker's keyboard probably (so you have all meta keys, arrow keys and more in the keyboard at once).

That's really great! i'm looking  forward to seeing it again someday :)

**Comment by Gottarocket** [comment]
Any updates on this . Missing those arrow keys

**Comment by patrickgold** [comment]
Since I did the layout rework the way the selection keyboard was implemented is not compatible anymore with the keyboard view. I can see if I can temporarily hack a solution to bring it to render again (the old XML/logic files for it are still there but just not in use), but can't promise if it will work. A better version of the selection keyboard fully compatible with the new layout engine will be implemented at a later stage.

**Comment by bege10** [comment]
For larger text manipulations I sometimes use this layout of Multiling O Keyboard which is fully customizable. I'd like to have something like that in FlorisBoard.
![screenshot_2021_09_08_15 05 54_1](https://user-images.githubusercontent.com/1184237/143706271-1fcf5534-485e-4122-92e0-1f64f5f2bc53.png)

**Comment by freMea** [comment]
I’m currently use the 0.4.0-alpha04 and I miss at least a selection mode. The selection keyboard (gboard like) was more than ok. I don’t know what is planned. There is this issue #1613 with a concept. As long as any integration bring back a selection mode, I’m happy.

**Comment by patrickgold** [comment]
Yeah I know a proper selection mode (without counting shift+arrow keys) is missing, and a selection mode will be integrated at some point, either in a dedicated selection keyboard or as the linked proposal or both. Atm I am focusing on word suggestions though, so I have little time for other features sadly.

**Comment by YakovL** [comment]
Really missing Selection Keyboard! As I've understood, it's [temporarily removed](https://github.com/florisboard/florisboard/issues/216#issuecomment-881971775), right? I think it's not useful to suggest details yet, is it? The one that's in Go Keyboard worked fine for me
